spec.bs

```diff
@@ -44,19 +58,81 @@ This specification extends the client-side storage available beyond cookies.
 
 <h2 id="extending-saa-to-non-cookie-storage">Extending SAA to non-cookie storage</h2>
 
-TBD
+This specification defines a method to request access to [=unpartitioned data=] beyond just cookies ({{Document/requestStorageAccess(types)}}), and a method to check if cookie access has specifically been granted ({{Document/hasUnpartitionedCookieAccess()}}).
 
-<h3 id="the-document-object">Changes to {{Document}}</h3>
+<div class=example>
 
-TBD
+Alex visits `https://social.example/`. The page sets a some local storage. This local storage has been set in a [=first-party-site context=].
 
-<h3 id="storage">Changes to various client-side storage mechanisms</h3>
+```javascript
+document.localStorage.setItem("userid", "1234");
+```
 
-TBD
+Later on, Alex visits `https://video.example/`, which has an <{iframe}> on it which loads `https://social.example/heart-button`. In this case, the `social.example` {{Document}} |doc| is in a [=third party context=], and the local storage set previously might or might not be visible depending on User Agent storage access policies.
 
-<h4 id="cookies">Cookies</h4>
+Script in the <{iframe}> can call |doc|`.`{{Document/requestStorageAccess(types)}} to request access.
 
-TBD
+```javascript
+let handle = await document.requestStorageAccess({localStorage: true});
+let userid = handle.localStorage.getItem("userid");
+```
+
+</div>
+
+<h3 id="document-changes">Changes to {{Document}}</h3>
+
+<pre class="idl">
+dictionary StorageAccessTypes {
+  boolean all = false;
+  boolean cookies = false;
+};
+
+interface StorageAccessHandle {
+};
+
+partial interface Document {
+  Promise&lt;boolean> hasUnpartitionedCookieAccess();
+  Promise&lt;StorageAccessHandle> requestStorageAccess(StorageAccessTypes types);
+};
+</pre>
+
+When invoked on {{Document}} |doc|, the <dfn export method for=Document><code>hasUnpartitionedCookieAccess()</code></dfn> method must run these steps:
+
+1. Invoke {{Document/hasStorageAccess()}}
+
+When invoked on {{Document}} |doc|, the <dfn export method for=Document><code>requestStorageAccess(types)</code></dfn> method must run these steps:
+
+1. Let |p| be [=a new promise=].
+1. If |types|'s member |all| is `false` and |types|'s member |cookies| is `false`, then [=/reject=] |p| with an "{{InvalidStateError}}" {{DOMException}} and return |p|.
+1. Let |requestUnpartitionedCookieAccess| be true if |types|'s member |all| or |types|'s member |cookies| is `true` and `false` otherwise.
+1. [=React=] |p| to the invocation of {{Document/requestStorageAccess()}} given |requestUnpartitionedCookieAccess| and invoke the following steps if the promise is fufilled (if the promise is rejected then apply that same rejection to |p|):
```

Comment:
I feel like just using the same algorithms as the SAA spec is a bit cleaner than using `react` here, any reason you decided to use that instead?